### PR TITLE
Feature/3196/update cw ltool

### PR DESCRIPTION
**Pull request by garyluu**
For #3866 

Using the version Michael Crusoe recommended. The newer version of ruamel.yaml throws error for duplicate keys (as it should).  Some tests in the CLI and here will have to be modified. 

See https://github.com/dockstore/dockstore-cli/pull/48 for passing tests

The plan is to:
- merge this
- create release
- modify and use it in the CLI PR
- test on tooltester
- revert both PRs if there's major problems
- make additional PRs if there's minor problems
- make adjustments + hotfix PRs if there's little to no problems

### CODE REVIEW

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage increased per file
===========================
+ dockstore-webservice/src/main/java/io/dockstore/webservice/DockstoreWebserviceApplication.java  14
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/dockerhub/DockerHubImage.java  24
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/Image.java  13
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/SourceCodeRepoFactory.java  2
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/BitBucketSourceCodeRepo.java  70
+ dockstore-webservice/src/main/java/io/dockstore/webservice/languages/NextflowHandler.java  61
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/EntryVersionHelper.java  19
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/BioWorkflow.java  8
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AliasResource.java  11
+ dockstore-webservice/src/main/java/io/dockstore/webservice/languages/WDLHandler.java  10
+ dockstore-webservice/src/main/java/io/swagger/api/impl/ToolsImplCommon.java  10
+ dockstore-webservice/src/main/java/io/dockstore/webservice/languages/CWLHandler.java  45
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/SourceCodeRepoInterface.java  1
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/Checksum.java  5
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/VersionDAO.java  12
+ dockstore-webservice/src/main/java/io/dockstore/webservice/languages/LanguageHandlerInterface.java  26
+ dockstore-webservice/src/main/java/io/dockstore/webservice/languages/LanguageHandlerFactory.java  31
+ dockstore-webservice/src/main/java/io/dockstore/webservice/languages/LanguagePluginHandler.java  7
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/EntryLabelHelper.java  5
+ dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/WorkflowVersionDAO.java  34
+ dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/AliasHelper.java  82
+ dockstore-webservice/src/main/java/io/dockstore/webservice/core/dockerhub/Results.java  11
+ dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java  11
         

Coverage decreased per file
===========================
- dockstore-webservice/src/main/java/io/dockstore/webservice/api/PrivilegeRequest.java  -100
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/database/EntryLite.java  -100
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/database/WorkflowPath.java  -100
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/Entry.java  -3
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/database/RSSToolPath.java  -100
- dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java  -7
- dockstore-common/src/main/java/io/dockstore/common/yaml/Filters.java  -100
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java  -44
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/rss/RSSWriter.java  -100
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/HostedWorkflowResource.java  -18
- dockstore-webservice/src/main/java/io/dockstore/webservice/api/Config.java  -100
- dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/SharedWorkflows.java  -100
- dockstore-common/src/main/java/io/dockstore/common/yaml/DockstoreYaml10.java  -50
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/UserResourceDockerRegistries.java  -78
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/StarrableResourceInterface.java  -100
- dockstore-webservice/src/main/java/io/swagger/api/impl/ToolsApiServiceImpl.java  -28
- dockstore-common/src/main/java/io/dockstore/common/DescriptorLanguage.java  -6
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/LambdaEventResource.java  -47
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/HostedToolResource.java  -60
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ZipSourceFileHelper.java  -66
- dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EventDAO.java  -64
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/CollectionEntry.java  -68
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubSourceCodeRepo.java  -33
- dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/Permission.java  -63
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/EntryUpdateTime.java  -55
- dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/TokenDAO.java  -24
- dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/PermissionsFactory.java  -20
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/EventResource.java  -70
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/Version.java  -3
- dockstore-webservice/src/main/java/io/swagger/model/ToolDescriptor.java  -44
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/Organization.java  -76
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/User.java  -21
- dockstore-common/src/main/java/io/dockstore/common/yaml/YamlWorkflow.java  -95
- dockstore-webservice/src/main/java/io/swagger/model/ToolTestsV1.java  -25
- dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/NotificationDAO.java  -85
- dockstore-common/src/main/java/io/dockstore/common/yaml/Service12.java  -100
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/TokenType.java  -7
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/EntryStarredSerializer.java  -34
- dockstore-webservice/src/main/java/io/dockstore/webservice/languages/ServicePrototypePlugin.java  -56
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ImageRegistryFactory.java  -47
- dockstore-webservice/src/main/java/io/swagger/api/ToolsApiV1.java  -89
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/EntryVersion.java  -42
- dockstore-webservice/src/main/java/io/openapi/api/impl/ToolsApiServiceImpl.java  -7
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/statelisteners/RSSListener.java  -30
- dockstore-common/src/main/java/io/dockstore/common/Registry.java  -11
- dockstore-webservice/src/main/java/io/swagger/api/impl/ToolClassesApiServiceImpl.java  -70
- dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/PermissionsInterface.java  -79
- dockstore-webservice/src/main/java/io/dockstore/webservice/CustomWebApplicationException.java  -25
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/AbstractImageRegistry.java  -7
- dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/NoOpPermissionsImpl.java  -10
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/rss/RSSEntry.java  -79
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractHostedEntryResource.java  -15
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/DockerRepoResource.java  -32
- dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/ToolDAO.java  -8
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/CacheConfigManager.java  -45
- dockstore-common/src/main/java/io/dockstore/common/yaml/Dockstore12Validator.java  -100
- dockstore-webservice/src/main/java/io/swagger/api/impl/MetadataApiServiceImpl.java  -89
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/MetadataResource.java  -64
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/OrganizationResource.java  -80
- dockstore-webservice/src/main/java/io/dockstore/webservice/Constants.java  -100
- dockstore-common/src/main/java/io/dockstore/common/yaml/AbstractYamlService.java  -100
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/OrganizationUpdateTime.java  -57
- dockstore-webservice/src/main/java/io/dockstore/webservice/DockstoreWebserviceConfiguration.java  -1
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/ExtendedUserData.java  -72
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/rss/RSSFeed.java  -100
- dockstore-common/src/main/java/io/dockstore/common/yaml/YamlService11.java  -100
- dockstore-webservice/src/main/java/io/openapi/api/impl/ToolClassesApiServiceImpl.java  -45
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/EntryResource.java  -5
- dockstore-webservice/src/main/java/io/swagger/api/impl/ApiV1VersionConverter.java  -86
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/Collection.java  -91
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/proposedGA4GH/ToolsExtendedApi.java  -33
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/CollectionResource.java  -72
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHelper.java  -85
- dockstore-webservice/src/main/java/io/dockstore/webservice/SimpleAuthorizer.java  -50
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/LambdaEvent.java  -84
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/Service.java  -28
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/statelisteners/SitemapListener.java  -25
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ConfigHelper.java  -55
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/statelisteners/ElasticListener.java  -18
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/statelisteners/TRSListener.java  -17
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/SourceFile.java  -8
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AliasableResourceInterface.java  -18
- dockstore-common/src/main/java/io/dockstore/common/DescriptorLanguageSubclass.java  -6
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/DockerRepoTagResource.java  -26
- dockstore-common/src/main/java/io/dockstore/common/Utilities.java  -17
- dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/UserDAO.java  -47
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/Notification.java  -54
- dockstore-common/src/main/java/io/dockstore/common/yaml/DockstoreYamlHelper.java  -85
- dockstore-webservice/src/main/java/io/swagger/model/ExtendedFileWrapper.java  -8
- dockstore-webservice/src/main/java/io/swagger/model/MetadataV1.java  -48
- dockstore-common/src/main/java/io/dockstore/common/yaml/DockstoreYaml11.java  -100
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/WorkflowVersion.java  -12
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/SourceControlConverter.java  -28
- dockstore-webservice/src/main/java/io/swagger/model/ToolDockerfile.java  -23
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/proposedGA4GH/ToolsApiExtendedServiceImpl.java  -50
- dockstore-webservice/src/main/java/io/swagger/api/MetadataApiV1.java  -34
- dockstore-common/src/main/java/io/dockstore/common/yaml/DockstoreYaml12.java  -100
- dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/OrganizationDAO.java  -83
- dockstore-webservice/src/main/java/io/swagger/model/ToolVersionV1.java  -52
- dockstore-webservice/src/main/java/io/swagger/api/impl/ApiV2BetaVersionConverter.java  -8
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/NotificationResource.java  -46
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/Token.java  -12
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/Tag.java  -3
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/ManualRegistry.java  -50
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/Tool.java  -14
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/database/RSSWorkflowPath.java  -100
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/PublicStateManager.java  -15
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/QuayImageRegistry.java  -1
- dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/ServiceDAO.java  -40
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/CollectionOrganization.java  -100
- dockstore-webservice/src/main/java/io/dockstore/webservice/api/Limits.java  -100
- dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/BioWorkflowDAO.java  -40
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/UserResource.java  -50
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/MetadataResourceHelper.java  -25
- dockstore-webservice/src/main/java/io/swagger/api/ToolClassesApiV1.java  -34
- dockstore-webservice/src/main/java/io/dockstore/webservice/resources/rss/RSSHeader.java  -81
- dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/LambdaEventDAO.java  -65
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitLabSourceCodeRepo.java  -3
- dockstore-webservice/src/main/java/io/dockstore/webservice/api/StarRequest.java  -62
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/database/ToolPath.java  -100
- dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubHelper.java  -26
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/OrganizationUser.java  -100
- dockstore-webservice/src/main/java/io/dockstore/webservice/permissions/InMemoryPermissionsImpl.java  -61
- dockstore-webservice/src/main/java/io/swagger/model/ToolV1.java  -60
- dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/WorkflowDAO.java  -37
- dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/CollectionDAO.java  -59
- dockstore-webservice/src/main/java/io/openapi/api/impl/ServiceInfoApiServiceImpl.java  -94
- dockstore-webservice/src/main/java/io/dockstore/webservice/core/Event.java  -25
         

Complexity increasing per file
==============================
- dockstore-common/src/main/java/io/dockstore/common/PipHelper.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dockstore/dockstore/pullRequest?prid=6341004&bid=20597263)

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- dockstore-common/src/main/java/io/dockstore/common/PipHelper.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/dockstore/dockstore/pullRequest?prid=6341003&bid=20597262)

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- dockstore-common/src/main/java/io/dockstore/common/PipHelper.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/dockstore/dockstore/pullRequest?prid=6341003&bid=20597262)

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- dockstore-common/src/main/java/io/dockstore/common/PipHelper.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dockstore/dockstore/pullRequest?prid=6341004&bid=20597263)

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- dockstore-common/src/main/java/io/dockstore/common/PipHelper.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dockstore/dockstore/pullRequest?prid=6341004&bid=20597263)

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- dockstore-common/src/main/java/io/dockstore/common/PipHelper.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/dockstore/dockstore/pullRequest?prid=6341003&bid=20597262)

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- dockstore-common/src/main/java/io/dockstore/common/PipHelper.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dockstore/dockstore/pullRequest?prid=6341004&bid=20597263)

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- dockstore-common/src/main/java/io/dockstore/common/PipHelper.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/dockstore/dockstore/pullRequest?prid=6341003&bid=20597262)

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- dockstore-common/src/main/java/io/dockstore/common/PipHelper.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dockstore/dockstore/pullRequest?prid=6341004&bid=20597263)

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- dockstore-common/src/main/java/io/dockstore/common/PipHelper.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/dockstore/dockstore/pullRequest?prid=6341003&bid=20597262)

File: scripts/install-tests.sh [hunk 1]
```diff
@@ -14,7 +14,7 @@ fi
 if [ "${TESTING_PROFILE}" = "regression-integration-tests" ]; then
     pip3 install --user -r dockstore-webservice/src/main/resources/requirements/1.6.0/requirements3.txt
```
Comment (denis-yuen) [inline scripts/install-tests.sh:15]: Tangent: https://github.com/dockstore/dockstore/issues/3880

**Review by denis-yuen — APPROVED** [review]
Pending passing tests

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- dockstore-common/src/main/java/io/dockstore/common/PipHelper.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/dockstore/dockstore/pullRequest?prid=6341003&bid=20597262)

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- dockstore-common/src/main/java/io/dockstore/common/PipHelper.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dockstore/dockstore/pullRequest?prid=6341004&bid=20597263)

**Review by NatalieEO — APPROVED** [review]

File: dockstore-common/src/test/java/io/dockstore/common/PipHelperTest.java [hunk 2]
```diff
@@ -10,8 +10,8 @@
 public class PipHelperTest {
     @Test
     public void convertSemVerToAvailableVersion() {
-        Assert.assertEquals("1.7.0", PipHelper.convertSemVerToAvailableVersion(PipHelper.DEV_SEM_VER));
-        Assert.assertEquals("1.7.0", PipHelper.convertSemVerToAvailableVersion(null));
+        Assert.assertEquals("1.10.0", PipHelper.convertSemVerToAvailableVersion(PipHelper.DEV_SEM_VER));
+        Assert.assertEquals("1.10.0", PipHelper.convertSemVerToAvailableVersion(null));
```
Comment (coverbeck) [inline dockstore-common/src/test/java/io/dockstore/common/PipHelperTest.java:14]: Ideally there would be tests for `PipHelper.convertSemVerToAvailableVersion("1.10.0")` and `PipHelper.convertSemVerToAvailableVersion("1.10.1)` if I understand the possible parameters that can be passed in.

**Review by coverbeck — APPROVED** [review]

File: dockstore-common/src/test/java/io/dockstore/common/PipHelperTest.java (hunk 2 above)
Comment (garyluu) [reply dockstore-common/src/test/java/io/dockstore/common/PipHelperTest.java:14]: Well, it's a good thing you mentioned this. A test failed

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- dockstore-common/src/main/java/io/dockstore/common/PipHelper.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/dockstore/dockstore/pullRequest?prid=6341003&bid=20597262)

**Comment by denis-yuen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- dockstore-common/src/main/java/io/dockstore/common/PipHelper.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dockstore/dockstore/pullRequest?prid=6341004&bid=20597263)

**Review by wshands — APPROVED** [review]

**Pull request merged**